core/src/main/java/com/linecorp/armeria/client/endpoint/SlowStartAwareEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,175 @@
+        this.numberOfSteps = numberOfSteps;
+    }
+
+    @Override
+    public EndpointSelector newSelector(EndpointGroup endpointGroup) {
+        return new SlowStartAwareWeightedRoundRobinSelector(endpointGroup, endpointWeightTransition);
+    }
+
+    /**
+     * An {@link EndpointSelector} that is similar to {@link WeightedRoundRobinStrategy} but it
+     * increases each {@link Endpoint} weight gradually.
+     */
+    private final class SlowStartAwareWeightedRoundRobinSelector implements EndpointSelector {
+        private final EndpointGroup endpointGroup;
+        private final AtomicInteger sequence = new AtomicInteger();
+        private volatile EndpointsAndWeights endpointsAndWeights;
+        private final EndpointWeightTransition endpointWeightTransition;
+
+        private volatile ScheduledFuture<?> updateEndpointWeightFuture;
+        private final Map<Endpoint, EndpointAndStep> currentWeights = new ConcurrentHashMap<>();
+
+        SlowStartAwareWeightedRoundRobinSelector(EndpointGroup endpointGroup,
+                                                 EndpointWeightTransition endpointWeightTransition) {
+            this.endpointGroup = endpointGroup;
+            this.endpointWeightTransition = endpointWeightTransition;
+            final List<Endpoint> endpoints = endpointGroup.endpoints();
+            endpointsAndWeights = new EndpointsAndWeights(endpoints);
+            endpointGroup.addListener(this::scheduleUpdateRequestWeight);
+        }
+
+        @Override
+        public EndpointGroup group() {
+            return endpointGroup;
+        }
+
+        @Override
+        public EndpointSelectionStrategy strategy() {
+            return SlowStartAwareEndpointSelectionStrategy.this;
+        }
+
+        @Override
+        public Endpoint select(ClientRequestContext ctx) {
+            final int currentSequence = sequence.getAndIncrement();
+            return endpointsAndWeights.selectEndpoint(currentSequence);
+        }
+
+        private void scheduleUpdateRequestWeight(List<Endpoint> newEndpoints) {
+            if (updateEndpointWeightFuture != null && !updateEndpointWeightFuture.cancel(false)) {
+                updateEndpointWeightFuture = null;
+            }
+            newEndpoints.forEach(e -> currentWeights.put(e, new EndpointAndStep(e)));
+            updateEndpointWeightFuture =
+                    executorService.scheduleAtFixedRate(() -> updateEndpointWeight(newEndpoints),
+                                                        0,
+                                                        slowStartInterval.toMillis(),
+                                                        TimeUnit.MILLISECONDS);
+        }
+
+        private void updateEndpointWeight(List<Endpoint> endpoints) {
+            endpointsAndWeights = new EndpointsAndWeights(
+                    endpoints.stream()
+                             .map(this::updateWeight)
+                             .collect(toImmutableList()));
```

Comment:
We should add logic to cancel the updater when all the weights have finished ramping, it's wasteful to run this forever in an environment with relatively stable endpoints.